lib/Authentication.ts

```diff
@@ -285,25 +320,39 @@ export default class Authentication {
     }
 
     // perhaps a more intellegent key choosing algorithm could be implemented here
+    // TODO get the key based on kid
     const documentKey = document.publicKey[0];
 
     const publicKey = this.factory.constructPublicKey(documentKey);
 
-    return this.signThenEncryptInternal(requesterNonce, privateKey, publicKey, content, accessToken);
+    return this.signThenEncryptInternal(requesterNonce, publicKey, content, accessToken);
   }
 
   /**
    * Given a JWE, retrieves the PrivateKey to be used for decryption
    * @param jweToken The JWE to inspect
    * @returns The PrivateKey corresponding to the JWE's encryption
    */
-  private getPrivateKeyForJwe (jweToken: JweToken): PrivateKey {
+  private async getPrivateKeyForJwe (jweToken: JweToken): Promise<string> {
     const keyId = jweToken.getHeader().kid;
-    const key = this.keys[keyId];
-    if (!key) {
-      throw new Error(`Unable to decrypt request; encryption key '${keyId}' not found`);
+    if (this.keys) {
+      const key = this.keys[keyId];

```

Comment:
Wouldn't this fail if the Jwe was encrypted using a private key only available in the keyReference, but with `this.keys` defined?

Reply:
We should not mix by value and by reference. I will add a check